File: cs/src/Connections/Messages/PortRelayRequestMessage.cs
// <copyright file="PortRelayRequestMessage.cs" company="Microsoft">
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license.
// </copyright>

using System;
using System.Text;
using Microsoft.DevTunnels.Ssh.Messages;
using Microsoft.DevTunnels.Ssh.IO;

namespace Microsoft.DevTunnels.Connections.Messages;

/// <summary>
/// Extends port-forward request messagse to include additional properties required
/// by the tunnel relay.
/// </summary>
public class PortRelayRequestMessage : PortForwardRequestMessage
{
    /// <summary>
    /// Access token with 'host' scope used to authorize the port-forward request.
    /// </summary>
    /// <remarks>
    /// A long-running host may need to handle the
    /// <see cref="ITunnelHost.RefreshingTunnelAccessToken" /> event to refresh the access token
    /// before forwarding additional ports.
    /// </remarks>
    public string? AccessToken { get; set; }

    /// <inheritdoc/>
    protected override void OnWrite(ref SshDataWriter writer)
    {
        base.OnWrite(ref writer);

        writer.Write(
            AccessToken ?? throw new InvalidOperationException("An access token is required."),
            Encoding.UTF8);
    }

    /// <inheritdoc/>
    protected override void OnRead(ref SshDataReader reader)
    {
        base.OnRead(ref reader);

        AccessToken = reader.ReadString(Encoding.UTF8);
    }
}

File: ts/src/connections/messages/portRelayRequestMessage.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { SshDataReader, SshDataWriter } from "@microsoft/dev-tunnels-ssh";
import { PortForwardRequestMessage } from "@microsoft/dev-tunnels-ssh-tcp";

/**
 * Extends port-forward request messagse to include additional properties required
 * by the tunnel relay.
 */
export class PortRelayRequestMessage extends PortForwardRequestMessage {
    /**
     * Access token with 'host' scope used to authorize the port-forward request.
     * A long-running host may need to handle the `refreshingTunnelAccessToken` event to
     * refresh the access token before forwarding additional ports.
     */
    public accessToken?: string;

    protected onWrite(writer: SshDataWriter): void {
        super.onWrite(writer);

        if (!this.accessToken) {
            throw new Error("An access token is required.");
        }

        writer.writeString(this.accessToken, 'utf8');
    }

    protected onRead(reader: SshDataReader): void {
        super.onRead(reader);

        this.accessToken = reader.readString('utf8');
    }
}
